### PR TITLE
New code/preview switcher

### PR DESCRIPTION
**Pull request by 1egoman**
Implements the new code/preview switcher interface in figma.

Screenshot 2024-10-15 at 4 56 26 PM
Screenshot 2024-10-15 at 4 56 37 PM

### CODE REVIEW

File: packages/web/src/components/apps/sidebar.tsx [hunk 1]
```diff
@@ -43,7 +44,25 @@ export default function Sidebar() {
...
+    if (tab === 'preview' && panel !== null) {
+      setSidePanelPreviousValue(panel);
+      _setPanel(null);
+    }
+  }, [tab, panel, sidePanelPreviousValue]);
+
   function setPanel(nextPanel: PanelType) {
+    switchTab('code');
     _setPanel(nextPanel === panel ? null : nextPanel);
   }
```
Comment (1egoman) [inline packages/web/src/components/apps/sidebar.tsx:67]: This behavior, through experimentation, seemed like it made sense. Effectively reopen the previously opened sidebar panel when a user goes back to "code" from "preview".

File: packages/web/src/components/apps/workspace/editor/editor.tsx [hunk 2]
```diff
@@ -9,25 +9,44 @@ import { useFiles } from '../../use-files';
...
+          ) : (
+            <div className="h-full flex items-center justify-center text-tertiary-foreground">
+              Use the file explorer to open a file for editing
+            </div>
+          )}
+        </div>
+      ) : null}
+
+      {/*
+      NOTE: applying hidden conditional like this keeps the iframe from getting mounted/unmounted
```
Comment (benjreinhart) [inline packages/web/src/components/apps/workspace/editor/editor.tsx:45]: 💯

**Review by benjreinhart — APPROVED** [review]
LGTM

File: packages/web/src/components/apps/use-preview.tsx [hunk 3]
```diff
@@ -42,6 +43,11 @@ export function PreviewProvider({ channel, children }: ProviderPropsType) {
     channel.push('preview:stop', {});
   }
 
+  // When the page initially loads, start the vite server
```
[the remark this answers is not in the document]
Comment (benjreinhart) [reply packages/web/src/components/apps/use-preview.tsx:46]: Probably but he can likely do that when working on some of the dependency related ticekts.

File: packages/web/src/components/apps/use-header-tab.tsx [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import React, { createContext, useContext, useState } from 'react';
```
[the remark this answers is not in the document]
Comment (benjreinhart) [reply packages/web/src/components/apps/use-header-tab.tsx:1]: Largely agree with this, especially if it's really only one value used in one or two places. That said, also not too concerned about abuse just yet, and we've had a different probably in the past where the top level component has so many state variables, effects, etc. which also become unwieldy. 

Fine to go either way here.

[the remark this answers is not in the document]
Comment (nichochar) [reply packages/web/src/components/apps/use-header-tab.tsx:1]: merging without changes is best then. Thanks+!

**Pull request merged**